src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,480 @@
+            this.logger.trace(
+                `Not checking key backup for session ${megolmSessionId} as it was already requested recently`,
+            );
+            return;
+        }
+
+        // We always add the request to the queue, even if we have a configuration problem (can't access backup).
+        // This is to make sure that if the configuration problem is resolved, we will try to download the key.
+        // This will happen after an initial sync, at this point the backup will not yet be trusted and the decryption
+        // key will not be available, but it will be just after the verification.
+        // We don't need to persist it because currently on refresh the sdk will retry to decrypt the messages in error.
+        this.queuedRequests.push({ roomId, megolmSessionId });
+
+        // Start the download loop if it's not already running.
+        this.downloadKeysLoop();
+    }
+
+    public stop(): void {
+        this.stopped = true;
+        this.backupManager.off(CryptoEvent.KeyBackupStatus, this.onBackupStatusChanged);
+        this.backupManager.off(CryptoEvent.KeyBackupFailed, this.onBackupStatusChanged);
+        this.backupManager.off(CryptoEvent.KeyBackupDecryptionKeyCached, this.onBackupStatusChanged);
+    }
+
+    /**
+     * Called when the backup status changes (CryptoEvents)
+     * This will trigger a check of the backup configuration.
+     */
+    private onBackupStatusChanged = (): void => {
+        // we want to check configuration
+        this.hasConfigurationProblem = false;
+        this.configuration = null;
+        this.getOrCreateBackupConfiguration(true).then((decryptor) => {
+            if (decryptor) {
+                // restart the download loop if it was stopped
+                this.downloadKeysLoop();
+            }
+        });
+    };
+
+    /** Returns true if the megolm session is already queued for download. */
+    private isAlreadyInQueue(roomId: string, megolmSessionId: string): boolean {
+        return this.queuedRequests.some((info) => {
+            return info.roomId == roomId && info.megolmSessionId == megolmSessionId;
+        });
+    }
+
+    /**
+     * Marks the session as not found in backup, to avoid retrying to soon for a key not in backup
+     * @param megolmSessionId - The megolm session ID that is missing.
+     * */
+    private markAsNotFoundInBackup(megolmSessionId: string): void {
+        const now = Date.now();
+        this.sessionLastCheckAttemptedTime.set(megolmSessionId, now);
+        // if too big make some cleaning to keep under control
+        if (this.sessionLastCheckAttemptedTime.size > 100) {
+            this.sessionLastCheckAttemptedTime = new Map(
+                Array.from(this.sessionLastCheckAttemptedTime).filter((sid, ts) => {
+                    return Math.max(now - ts, 0) < this.backoffDuration;
+                }),
+            );
```

Comment:
We might end up doing this for every call to markAsNotFoundInBackup (since there is no guarantee that removing the old entries will bring the size below 100), so I'm a bit worried about performance here, too.

But again, happy to leave this for now if you prefer.

Reply:
I also have concerns, more about the fact that it could only grow and get out of control (particularly with the current relation handling that fetches a lot of events) rather than performance but indead valid. Let's iterate